### ISSUE TITLE
Optimizing hashing performance

### ISSUE THREAD

**Issue by fabiospampinato** [nodejs/performance#136]
### What is the problem this feature will solve?

Making the hash functions significantly faster.

### What is the feature you are proposing to solve the problem?

I'm not sure what the best option for this is, but running the following file:

```ts
import crypto from 'node:crypto';

const value = '/Users/fabio/path/to/some/random/file.js';

console.time ( 'crypto' );
for ( let i = 0; i < 100_000; i++ ) {
  crypto.createHash("sha1").update(value).digest("hex");
}
console.timeEnd ( 'crypto' );
```

I see the following output: 

```ts
~ ❯ node hashbench.js
crypto: 106.557ms
~ ❯ node hashbench.js
crypto: 108.052ms
~ ❯ node hashbench.js
crypto: 107.642ms
~ ❯ node hashbench.js
crypto: 107.136ms
~ ❯ node hashbench.js
crypto: 107.747ms
~ ❯ bun hashbench.js
[29.98ms] crypto
~ ❯ bun hashbench.js
[33.87ms] crypto
~ ❯ bun hashbench.js
[30.74ms] crypto
~ ❯ bun hashbench.js
[30.78ms] crypto
~ ❯ bun hashbench.js
[30.12ms] crypto
```

Basically Node's sha1 function seems at least 3x slower than Bun's.

Hashing is at the core of many important things, so I'd argue it's important to hash as fast as possible since that would speed up a cascade of use cases.

I'm not sure what the best solution is here, but if Bun is 3x faster than Node here presumably there's a lot of room for improvement.

**Comment by H4ad** [comment]
I spent a few minutes trying to investigate this and here are some numbers.

### Bun

> [39.36ms] crypto

### Node.js 16.20.1

> crypto: 85.105ms

### Node.js 18.18.1

> 105.906ms

### Node.js 20.9.0

> crypto: 91.476ms

### Node.js 21.1.0

> crypto: 90.148ms

---

About bun, the `update` function is handled by [CryptoDigest.cpp](https://github.com/oven-sh/bun/blob/22818e6806cfff66804a74bdbd71f90d195f8183/src/bun.js/bindings/webcrypto/CryptoDigest.cpp).

Apparently, `bun` uses openssl 1.x, Node.js 16.x also uses that same version, after 16.x Node.js started using openssl 3.x

Instead of [EVP_* functions](https://www.openssl.org/docs/manmaster/man3/EVP_DigestUpdate.html) ([code](https://github.com/nodejs/node/blob/ed5cb37ccbc44e2310286db309edc521d8deb91d/src/crypto/crypto_hash.cc#L139-L143)), `bun` uses [SHA1 Functions](https://www.openssl.org/docs/man3.0/man3/SHA1_Update.html) ([code](https://github.com/oven-sh/bun/blob/22818e6806cfff66804a74bdbd71f90d195f8183/src/bun.js/bindings/webcrypto/CryptoDigest.cpp#L32-L37)).

I updated the benchmark to:

```js
const crypto = require('node:crypto');

const value = '/Users/fabio/path/to/some/random/file.js';
const hash = crypto.createHash("sha1");

const start = performance.now();
hash.update(value)
console.log(`End: ${performance.now() - start}`);
```

Bun: 0.0250780000000006ms
Node.js 21.1.0: 0.03838299959897995ms
Node.js 21.1.0 (with `%OptimizeFunctionOnNextCall(hash.update)`): 0.03241100162267685ms

So, even calling only `update` is slower than `bun`, so maybe using `SHA1_UPDATE` instead of `EVP_DigestUpdate` can help.

**Comment by Uzlopak** [comment]
But is is even recommendable to use non EVP Calls?

**Comment by Uzlopak** [comment]
Btw. It is crazy how slow it is  to sha256 hmac a string. It is something i cant optimize in probot and results in about 6k req/s performance of a probot server.

**Comment by H4ad** [comment]
> But is is even recommendable to use non EVP Calls?

I don't have any idea, in the documentation page, they deprecated some overloads of `SHA1` functions but they still support it if you change the arguments to pass `context, data, length`.

So, I think is safe to use.

> Btw. It is crazy how slow it is to sha256 hmac a string. It is something i cant optimize in probot and results in about 6k req/s performance of a probot server.

Once I cached all the string and then performed only one `update` and `digest`, this improved the performance significantly.

**Comment by tniessen** [comment]
> Apparently, `bun` uses openssl 1.x, Node.js 16.x also uses that same version, after 16.x Node.js started using openssl 3.x

There are various known performance issues related to OpenSSL 3. It's possible that they negatively affect performance of hashing through `EVP_*` as well, but someone would have to benchmark the C/C++ code itself to be sure.

> Instead of [EVP_* functions](https://www.openssl.org/docs/manmaster/man3/EVP_DigestUpdate.html) ([code](https://github.com/nodejs/node/blob/ed5cb37ccbc44e2310286db309edc521d8deb91d/src/crypto/crypto_hash.cc#L139-L143)), `bun` uses [SHA1 Functions](https://www.openssl.org/docs/man3.0/man3/SHA1_Update.html) ([code](https://github.com/oven-sh/bun/blob/22818e6806cfff66804a74bdbd71f90d195f8183/src/bun.js/bindings/webcrypto/CryptoDigest.cpp#L32-L37)).

The `EVP_*` functions are recommended over low-level functions etc. The latter are gradually being deprecated and using them might break compatibility with OpenSSL providers/engines.

---

FWIW, I proposed `crypto.hash()` at some point (see https://github.com/nodejs/node/pull/42233), which would be faster than `crypto.Hash` for one-shot hashing. It might still be slower than an equivalent implementation using OpenSSL 1.1.1 and low-level OpenSSL APIs.

**Comment by ronag** [comment]
Are we using OpenSSL 3.0 or 3.1?

**Comment by Uzlopak** [comment]
@ronag 
A fork of 3.0 with quic patched. No original OpenSSL 3.0

We are basically cut off from direct OpenSSL updates.

**Comment by ronag** [comment]
So basically we are stuck with this until we update OpenSSL.

**Comment by ronag** [comment]
Anyone considered switching to BoringSSL?

**Comment by Uzlopak** [comment]
I guess a PR to switch to BoringSSL would result in a lot of negative feedback. Also BoringSSL does not guarantee stable ABI.

https://boringssl.googlesource.com/boringssl/+/HEAD/PORTING.md#porting-from-openssl-to-boringssl

> Note: BoringSSL does not have a stable API or ABI. It must be updated with its consumers. It is not suitable for, say, a system library in a traditional Linux distribution. For instance, Chromium statically links the specific revision of BoringSSL it was built against. Likewise, Android's system-internal copy of BoringSSL is not exposed by the NDK and must not be used by third-party applications.

E.g. @mcollina mentioned that stable ABI is important for him. So I assume for other it is also important.

**Comment by ronag** [comment]
> E.g. @mcollina mentioned that stable ABI is important for him. So I assume for other it is also important.

Is this a theoretical issue or a practical one? Do they actually break the ABI that often?

**Comment by targos** [comment]
https://boringssl.googlesource.com/boringssl/+/HEAD/README.md

> Although BoringSSL is an open source project, it is not intended for general use, as OpenSSL is. We don't recommend that third parties depend upon it. Doing so is likely to be frustrating because there are no guarantees of API or ABI stability.

**Comment by GeoffreyBooth** [comment]
@nodejs/performance @nodejs/crypto

**Comment by lemire** [comment]
Try the following... Start with this script... (call it `hash.mjs`)

```JavaScript

import crypto from 'node:crypto';

const value = '/Users/fabio/path/to/some/random/file.js';

console.time ( 'crypto' );
for ( let i = 0; i < 1000_000; i++ ) {
  crypto.createHash("sha1").update(value).digest("hex");
}
console.timeEnd ( 'crypto' );
```

(I picked the script as-is, but I extended to loop to 1000_000 repetitions.)

The conjecture is that the hashing is slow, and by that, presumably, we mean the `update` call. Ok. Let us trim the code... create a new script that does not update... (call it `bhash.msj`)


```JavaScript

import crypto from 'node:crypto';

console.time ( 'crypto' );
for ( let i = 0; i < 1000_000; i++ ) {
  crypto.createHash("sha1");
}
console.timeEnd ( 'crypto' );
```

If hashing done by update is the bottleneck, this second script should be very fast compared to the previous one. Let us see... I use Node.js v20.10.0.

```
$ node hash.mjs 
crypto: 1.442s
$ node bhash.mjs 
crypto: 1.039s
```

Yep. It is not hashing. Only 0.4 s are spent on hashing. The bulk of the time is spent on `crypto.createHash("sha1")`.

Let us try with Bun now...

```
$ bun hash.mjs 
[750.63ms] crypto
$ bun bhash.mjs 
[105.19ms] crypto
```

So bun spends 0.64 s on hashing... roughly as much as Node.js. The big difference between Node and Bun is that Bun is 10 times faster on the `crypto.createHash("sha1")` part.

So let us try something else...  let us call `crypto.createHash` once:

```JavaScript
import crypto from 'node:crypto';

const value = '/Users/fabio/path/to/some/random/file.js';

console.time ( 'crypto' );
const hasher = crypto.createHash("sha1");
for ( let i = 0; i < 1000_000; i++ ) {
  hasher.update(value);
}
console.timeEnd ( 'crypto' );
```

I get...
```
$ node chash.mjs 
crypto: 172.675ms
$ bun chash.mjs 
[170.99ms] crypto
```

Yep. So we have narrowed it down to `crypto.createHash`.


So let us profile the `bhash.mjs` script with Node:

```
   7.69%  node     libc.so.6             [.] pthread_rwlock_unlock@@GLIBC_2.34
   7.35%  node     libc.so.6             [.] malloc
   5.51%  node     libc.so.6             [.] pthread_rwlock_rdlock@GLIBC_2.2.5
   4.77%  node     node                  [.] Builtins_GetProperty
   4.65%  node     libc.so.6             [.] _int_free
   4.10%  node     libc.so.6             [.] _int_malloc
```

So we are spending time allocating memory and doing some threading. These are expensive jobs...

And now Bun...

```
   5.08%  bun          bun                   [.] 0x0000000002e52279
   4.26%  bun          bun                   [.] 0x0000000002e52897
   3.56%  bun          bun                   [.] 0x00000000030aabdf
   3.55%  bun          libc.so.6             [.] __memset_evex_unaligned_erms
   2.20%  bun          libc.so.6             [.] __strncasecmp_l_evex
```

My conclusion is that though the problem might come from OpenSSL, and hashing... it does not appear to be so. As you can see, just `Builtins_GetProperty` seems to be a bottleneck.

Maybe we should get the advice of someone who understands deeply the C++ architecture:  @joyeecheung

**Comment by joyeecheung** [comment]
Actually regarding the OpenSSL overhead the solution seems surprisingly simple - we are just under-initializing OpenSSL: https://github.com/nodejs/node/pull/51019

**Comment by joyeecheung** [comment]
Actually https://github.com/nodejs/performance/issues/136#issuecomment-1837269580 was incorrect, I was just looking at the numbers from the Oilpan prototype, initializing it with `OPENSSL_INIT_ADD_ALL_DIGESTS` won't make a differenct because `EVP_get_digestbyname` would also do that.

On another note, it may be a good idea to just move away from `EVP_get_digestbyname` if we are on OpenSSL 3. Locally switching to `EVP_MD_fetch` (but without proper lifecycle management so it leaks) can speed up `createHash()` by another 10% - though handling the life cycle of that ref-counted EVP_MD in the existing implementation can be somewhat tricky.

**Comment by Uzlopak** [comment]
Doesnt the OpenSSL docs say something about avoiding the lookup functions like EVP_get_digestbyname? I think I read something about this. Something along the lines that OpenSSL 3 is totally object oriented and thus e.g. you register digests in an Object at runtime. So calling something like EVP_get_digestbyname means basically, that OpenSSL goes through each name and checks if it finds the proper digest.

**Comment by joyeecheung** [comment]
Yes, though I think we'll need a bit of internal refactoring to move on the explicit fetching because so far most of the crypto classes don't think about freeing the EVP_MD.

**Comment by fabiospampinato** [comment]
> the other bottleneck is the string decoding part that converts input from a string to a buffer.

As I understand it V8 keeps track of whether a string is encoded in latin1 or not, because if it's using latin1 it can just allocate 1 byte per character rather than 2. In that case, for latin1 strings, if the rope is flattened and whatever else may be needed, shouldn't decoding be essentially free because the slice of memory that the string points to can just be hashed directly? Maybe it's more complicated than that 🤔

**Comment by joyeecheung** [comment]
> In that case, for latin1 strings, if the rope is flattened and whatever else may be needed, shouldn't decoding be essentially free because the slice of memory that the string points to can just be hashed directly?

V8 does not hand over the raw bytes to embedders, only a copy. Also it's only a pure copy for ASCII-only strings. For strings that are Latin1 but not ASCII there's still a bit of conversion to split each code point into two bytes.

**Comment by lemire** [comment]
@joyeecheung 

If you effectively made Node 2x faster… it feels like it is quite a success… it is not enough to beat Bun but it is enough so that it is no longer embarrassing for Node. Especially if you achieved that with just a few lines of code!

**Comment by joyeecheung** [comment]
I think "making Node 2x" faster isn't a very accurate description of the overall situation...or that Node.js being 3x slower than Bun already is a limited description of the situation - the 3x slowness is conditioned to:

1. This is hashing a lot (100K) of short strings (in a tight loop where GC performance matters, but this isn't as common in the real world)
2. The input is not streamed.

Once the input is not a string but a buffer, or the input is streamed, or there aren't as many as 100K inputs being hashed in the tight loop, the magnitude of slowness decreases. For example if the workload is just hashing 1000 buffers with 100 * 10000 chunks, then Node.js and Bun can already finish the workload in about the same time - with or without my changes. If they are 1000 strings in 100 * 10000 chunks, then Node.js is about 1.8x slower - with or without my changes.

And then, after getting rid of what can be done on the Node.js side, the primary bottleneck we found are GC (when hashing a lot of small input in a tight loop) and string -> buffer conversion (when hashing input streamed in many chunks). These are mostly V8 v.s. JSC differences, not Node.js v.s Bun per-se.

**Comment by lemire** [comment]
@joyeecheung Point taken. My comment was in the scope of the microbenchmark that motivates the issue. I did not mean to imply that you made Node 2x faster in general.

The string-to-buffer conversion issue is annoying.

**Comment by fabiospampinato** [comment]
It looks like somebody had a [similar idea](https://github.com/nodejs/node/pull/42233) at some point.

It'd be cool if kinda common APIs like that could be made 3x faster. Having both a `hash` and a `createHash` function seems a little weird.

**Comment by lemire** [comment]
> It'd be cool if kinda common APIs like that could be made 3x faster. Having both a hash and a createHash function seems a little weird though

I think you could maybe do it with a lazy/on-demand approach. So right now you compute... `crypto.createHash("sha1").update(value).digest("hex")` by eagerly doing all the work as soon as you can. This means that you need to have a non-trivial Hash object and do JavaScript-to-C++ calls repeatedly.

Instead, you could have a lightweight JavaScript object that simply collects the information without doing anything. Then when the has value is required, it computes it. In fact, if @joyeecheung implemented `createHash`, then an on-demand approach like what I describe could be just a few lines of code away, at least in prototypical form.

**Comment by joyeecheung** [comment]
> Instead, you could have a lightweight JavaScript object that simply collects the information without doing anything.

I assume this is talking about buffering in JS land (possibly with a JS array) - the overhead of iterating over that array might already be another bottleneck until we get the new Array iteration API in https://github.com/nodejs/node/pull/50115. It would also increase the memory overhead when there are more updates/larger inputs that won't be otherwise kept alive. For a non-streaming use case, a one-shot digest implementation might still fare better, or users can do the buffering themselves and pass the concatenated buffers into a one-shot digest implementation if for their use case the memory overhead of buffering the original data matter less than hashing performance.

**Comment by joyeecheung** [comment]
> Having both a hash and a createHash function seems a little weird though 🤔

The crypto interface is largely just mapping onto OpenSSL APIs, so I think a one-shot digest helper + a streaming API would still make sense as there are also similar interface parities in OpenSSL (`EVP_Digest` v.s. `EVP_DigestInit`/`EVP_DigestUpdate`/`EVP_DigestFinal_*`). They serve different purposes - a streaming API would fair better in an asynchronous application and has lower memory overhead, whereas the one-shot digest is faster if the use case is just computing a digest of a piece of immediately available data synchronously.

**Comment by lemire** [comment]
> I assume this is talking about buffering in JS land (possibly with a JS array)

What I am saying is that `crypto.createHash("sha1")`  only needs an object that contains `"sha1"`. It does not need to do anything, at least nothing in C++ land right away. I am just alluding to the design space you have.

I do agree with you, btw, that it makes a lot of sense to do a one-shot hashing... But what I am also saying is that `crypto.createHash("sha1").update(buffer).digest(...)` should not have wildly different performance characteristics.

**Comment by joyeecheung** [comment]
Yeah, there could be some micro-optimization that only starts a object-based implementation upon second update / when the first update contains a big enough input (say >1MB). Otherwise fallback to a one-shot digest implementation. Though the prerequisite would still be having a one-shot digest implementation...

**Comment by lemire** [comment]
> Yeah, there could be some micro-optimization that only starts a object-based implementation upon second update / when the first update contains a big enough input (say >1MB). Otherwise fallback to a one-shot digest implementation. Though the prerequisite would still be having a one-shot digest implementation..

Precisely. :heart: :heart: The idea is that if you propose a new function, not everyone will rush to adopt it. But it might be possible to provide similar benefits to people who use the conventional API. :dash:

**Comment by merceyz** [comment]
If you want a real world example where `createHash` is called a lot then Yarn might provide that.

Setup:

```sh
git clone https://github.com/yarnpkg/berry
cd berry
git checkout db6210f48355d2986e965f90009b22f18d3b6342
yarn build:cli --no-minify
```

Running the following command will call `createHash` 41932 times:
```sh
YARN_IGNORE_PATH=1 CI=true node ./packages/yarnpkg-cli/bundles/yarn.js
```

and the following will, at the time of writing, call it 73048 times:
```
rm yarn.lock
YARN_IGNORE_PATH=1 node ./packages/yarnpkg-cli/bundles/yarn.js
```

**Comment by joyeecheung** [comment]
I looked into the ["buffering input" ](https://github.com/joyeecheung/node/tree/buffer-digest) idea for `crypto.createHash` and it turns out to be more complicated than I thought - most importantly, errors would be delayed until the second `update()` or `digest()` is called (e.g. if you pass an algorithm that the OpenSSL version doesn't support, you can't get an error immediately, because we don't eagerly cache all the supported algorithms to avoid overhead for the first access). Added to it the fact that Hash objects are also streams, it could be difficult to get the error handling correctly, So I'll just go with https://github.com/nodejs/node/pull/51044 for now.

**Comment by H4ad** [comment]
Since https://github.com/nodejs/node/pull/51044 was merged, I think we can close this issue.

Any other discussion about OpenSSL and its perf, I think we should keep at https://github.com/nodejs/performance/issues/72

**Comment by larenztate** [comment]
They closed suddenly without notice while there's balance of 8000 dollars ! Now they don't allow trading or more than that, asking me to open a new account !! What about my money !!! I am currently working with ''wikihelpdesksma﹫ℊmaℹ️l .¢OM '' for assistance but i do not know if this can help me?
